File: app/solver/pow_solver.js
const fs = require('fs');
const path = require('path');

const wasmBuffer = fs.readFileSync(path.join(__dirname, 'sha3_wasm.wasm'));
let wasmInstance = null;

async function initWasm() {
  if (!wasmInstance) {
    const wasmModule = await WebAssembly.instantiate(wasmBuffer, { wbg: {} });
    wasmInstance = wasmModule.instance.exports;
  }
  return wasmInstance;
}

let WASM_VECTOR_LEN = 0;
const encoder = new TextEncoder();

function passStringToWasm(wasm, arg) {
  const buf = encoder.encode(arg);
  const ptr = wasm.__wbindgen_export_0(buf.length, 1) >>> 0;
  new Uint8Array(wasm.memory.buffer).subarray(ptr, ptr + buf.length).set(buf);
  WASM_VECTOR_LEN = buf.length;
  return ptr;
}

async function solve(algorithm, challenge, salt, difficulty, expireAt, signature, targetPath) {
  if (algorithm !== 'DeepSeekHashV1') {
    throw new Error('Unsupported algorithm: ' + algorithm);
  }
  const wasm = await initWasm();
  const prefix = `${salt}_${expireAt}_`;
  const retptr = wasm.__wbindgen_add_to_stack_pointer(-16);
  try {
    const ptr0 = passStringToWasm(wasm, challenge);
    const len0 = WASM_VECTOR_LEN;
    const ptr1 = passStringToWasm(wasm, prefix);
    const len1 = WASM_VECTOR_LEN;

    wasm.wasm_solve(retptr, ptr0, len0, ptr1, len1, difficulty);
    const dataView = new DataView(wasm.memory.buffer);
    const r0 = dataView.getInt32(retptr + 0, true);
    const r1 = dataView.getFloat64(retptr + 8, true);
    if (r0 === 0) return null;
    return r1;
  } finally {
    wasm.__wbindgen_add_to_stack_pointer(16);
  }
}

if (require.main === module) {
  const args = process.argv.slice(2);
  if (args.length >= 6) {
    const [algorithm, challenge, salt, difficulty, expireAt, signature, targetPath] = args;
    solve(algorithm, challenge, salt, parseInt(difficulty), parseInt(expireAt), signature, targetPath || '/api/v0/chat/completion')
      .then(answer => {
        const responseObj = {
          algorithm,
          challenge,
          salt,
          answer,
          signature,
          target_path: targetPath || '/api/v0/chat/completion'
        };
        const header = Buffer.from(JSON.stringify(responseObj)).toString('base64');
        console.log(JSON.stringify({ answer, header }));
      })
      .catch(err => {
        console.error(err);
        process.exit(1);
      });
  } else {
    console.error("Usage: node pow_solver.js <algorithm> <challenge> <salt> <difficulty> <expireAt> <signature> [targetPath]");
    process.exit(1);
  }
}

module.exports = { solve };

File: app/solver/pow_solver.py
import json
import subprocess
import asyncio
from pathlib import Path

POW_JS_PATH = Path(__file__).parent / "pow_solver.js"

def solve_pow_sync(algorithm: str, challenge: str, salt: str, difficulty: int, expire_at: int, signature: str, target_path: str = "/api/v0/chat/completion") -> dict:
    """
    Synchronously solve DeepSeek PoW challenge using the local WebAssembly engine.
    Returns dict with 'answer' and base64 'header'.
    """
    cmd = [
        "node",
        str(POW_JS_PATH),
        algorithm,
        challenge,
        salt,
        str(difficulty),
        str(expire_at),
        signature,
        target_path
    ]
    res = subprocess.run(cmd, capture_output=True, text=True, check=True)
    return json.loads(res.stdout.strip())

async def solve_pow_async(algorithm: str, challenge: str, salt: str, difficulty: int, expire_at: int, signature: str, target_path: str = "/api/v0/chat/completion") -> dict:
    """
    Asynchronously solve DeepSeek PoW challenge.
    """
    proc = await asyncio.create_subprocess_exec(
        "node",
        str(POW_JS_PATH),
        algorithm,
        challenge,
        salt,
        str(difficulty),
        str(expire_at),
        signature,
        target_path,
        stdout=asyncio.subprocess.PIPE,
        stderr=asyncio.subprocess.PIPE
    )
    stdout, stderr = await proc.communicate()
    if proc.returncode != 0:
        raise RuntimeError(f"PoW solve failed: {stderr.decode('utf-8')}")
    return json.loads(stdout.decode('utf-8').strip())
